File: mobile/lib/models/market_price.dart
class MarketPriceModel {
  final String id;
  final String state;
  final String district;
  final String market;
  final String commodity;
  final String variety;
  final String grade;
  final double minPrice;
  final double maxPrice;
  final double modalPrice;
  final String date;

  MarketPriceModel({
    required this.id,
    required this.state,
    required this.district,
    required this.market,
    required this.commodity,
    required this.variety,
    required this.grade,
    required this.minPrice,
    required this.maxPrice,
    required this.modalPrice,
    required this.date,
  });

  factory MarketPriceModel.fromJson(Map<String, dynamic> json) {
    return MarketPriceModel(
      id: json['id']?.toString() ?? '',
      state: json['state']?.toString() ?? 'N/A',
      district: json['district']?.toString() ?? 'N/A',
      market: json['market']?.toString() ?? 'N/A',
      commodity: json['commodity']?.toString() ?? 'N/A',
      variety: json['variety']?.toString() ?? 'Standard',
      grade: json['grade']?.toString() ?? 'FAQ',
      minPrice: (json['minPrice'] as num?)?.toDouble() ?? 0.0,
      maxPrice: (json['maxPrice'] as num?)?.toDouble() ?? 0.0,
      modalPrice: (json['modalPrice'] as num?)?.toDouble() ?? 0.0,
      date: json['date']?.toString() ?? '',
    );
  }
}

File: server/src/models/MarketPrice.js
const mongoose = require('mongoose');

const marketPriceSchema = new mongoose.Schema(
  {
    state: {
      type: String,
      required: [true, 'State is required'],
      trim: true,
      index: true,
    },
    district: {
      type: String,
      required: [true, 'District is required'],
      trim: true,
      index: true,
    },
    market: {
      type: String,
      required: [true, 'Market is required'],
      trim: true,
      index: true,
    },
    commodity: {
      type: String,
      required: [true, 'Commodity is required'],
      trim: true,
      index: true,
    },
    variety: {
      type: String,
      default: 'Other',
      trim: true,
    },
    grade: {
      type: String,
      default: 'FAQ',
      trim: true,
    },
    minPrice: {
      type: Number,
      required: [true, 'Minimum price is required'],
      min: [0, 'Minimum price cannot be negative'],
    },
    maxPrice: {
      type: Number,
      required: [true, 'Maximum price is required'],
      min: [0, 'Maximum price cannot be negative'],
    },
    modalPrice: {
      type: Number,
      required: [true, 'Modal price is required'],
      min: [0, 'Modal price cannot be negative'],
    },
    date: {
      type: Date,
      required: [true, 'Price date is required'],
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

// Compound unique index to prevent duplicate record insertion
marketPriceSchema.index(
  {
    state: 1,
    district: 1,
    market: 1,
    commodity: 1,
    variety: 1,
    grade: 1,
    date: 1,
  },
  { unique: true }
);

// Performance indexes for frequent queries
marketPriceSchema.index({ commodity: 1, date: -1 });
marketPriceSchema.index({ state: 1, commodity: 1, date: -1 });
marketPriceSchema.index({ market: 1, commodity: 1, date: -1 });

module.exports = mongoose.model('MarketPrice', marketPriceSchema);
